src/main/java/com/cloudbees/jenkins/plugins/sshagent/SSHAgentStepExecution.java

```diff
@@ -149,97 +115,17 @@ private void initRemoteAgent() throws IOException, InterruptedException {
-        listener.getLogger().println("[ssh-agent] Looking for ssh-agent implementation...");
-        Map<String, Throwable> faults = new LinkedHashMap<>();
-        for (RemoteAgentFactory factory : Jenkins.get().getExtensionList(RemoteAgentFactory.class)) {
-            if (factory.isSupported(launcher, listener)) {
-                try {
-                    listener.getLogger().println("[ssh-agent]   " + factory.getDisplayName());
-                    agent = factory.start(this, listener, tempDir(workspace));
-                    break;
-                } catch (Throwable t) {
-                    faults.put(factory.getDisplayName(), t);
-                }
-            }
-        }
-        if (agent == null) {
-            listener.getLogger().println("[ssh-agent] FATAL: Could not find a suitable ssh-agent provider");
-            listener.getLogger().println("[ssh-agent] Diagnostic report");
-            for (Map.Entry<String, Throwable> fault : faults.entrySet()) {
-                listener.getLogger().println("[ssh-agent] * " + fault.getKey());
-                StringWriter sw = new StringWriter();
-                fault.getValue().printStackTrace(new PrintWriter(sw));
-                for (String line : StringUtils.split(sw.toString(), "\n")) {
-                    listener.getLogger().println("[ssh-agent]     " + line);
-                }
-            }
-            throw new RuntimeException("[ssh-agent] Could not find a suitable ssh-agent provider.");
-        }
+        agent = new ExecRemoteAgent(launcher, listener);
 
         for (SSHUserPrivateKey userPrivateKey : userPrivateKeys) {
             final Secret passphrase = userPrivateKey.getPassphrase();
             final String effectivePassphrase = passphrase == null ? null : passphrase.getPlainText();
             for (String privateKey : userPrivateKey.getPrivateKeys()) {
-                agent.addIdentity(privateKey, effectivePassphrase, SSHAgentBuildWrapper.description(userPrivateKey),
-                        listener);
+                agent.addIdentity(privateKey, effectivePassphrase, SSHAgentBuildWrapper.description(userPrivateKey), workspace, launcher, listener);
             }
         }
 
         listener.getLogger().println(Messages.SSHAgentBuildWrapper_Started());
-        socket = agent.getSocket();
-        sockets.add(socket);
-    }
-
-    /**
-     * Shuts down the current SSH Agent and purges socket files.
-     */
-    private void cleanUp() throws Exception {
-        try {
-            TaskListener listener = getContext().get(TaskListener.class);
-            if (agent != null) {
-                agent.stop(listener);
-                listener.getLogger().println(Messages.SSHAgentBuildWrapper_Stopped());
-            }
-        } finally {
-            purgeSockets();
-        }
-    }
-
-    /**
-     * Purges all socket files created previously.
-     * Especially useful when Jenkins is restarted during the execution of this step.
```

Comment:
Something else dating to #4 but of unclear purpose. We already call `ssh-agent -k`, which deletes the socket file.